pennylane/utils.py

```diff
@@ -19,16 +19,93 @@
 from collections.abc import Iterable
 from collections import OrderedDict
 import copy
-import numbers
 import functools
 import inspect
+import itertools
+import numbers
+from operator import matmul
 
 import numpy as np
 
 import pennylane as qml
 from pennylane.variable import Variable
 
 
+def decompose_hamiltonian(H, hide_identity=False):
+    r"""Decomposes a Hermitian matrix into a linear combination of Pauli operators.
+
+    Args:
+        H (array[complex]): a Hermitian matrix of dimension :math:`2^n\times 2^n`
+        hide_identity (bool): does not include the :class:`~.Identity` observable within
+            the tensor products of the decomposition if ``True``
+
+    Returns:
+        tuple[list[float], list[~.Observable]]: a list of coefficients and a list
+        of corresponding tensor products of Pauli observables that decompose the Hamiltonian.
+
+    **Example:**
+
+    We can use this function to compute the Pauli operator decomposition of an arbitrary Hermitian
+    matrix:
+
+    >>> A = np.array(
+    ... [[-2, -2+1j, -2, -2], [-2-1j,  0,  0, -1], [-2,  0, -2, -1], [-2, -1, -1,  0]])
+    >>> coeffs, obs_list = decompose_hamiltonian(A)
+    >>> coeffs
+    [-1.0, -1.5, -0.5, -1.0, -1.5, -1.0, -0.5, 1.0, -0.5, -0.5]
+
+    We can use the output coefficients and tensor Pauli terms to construct a :class:`~.Hamiltonian`:
+
+    >>> H = qml.Hamiltonian(coeffs, obs_list)
+    >>> print(H)
+    (-1.0) [I0 I1]
+    + (-1.5) [X1]
+    + (-0.5) [Y1]
+    + (-1.0) [Z1]
+    + (-1.5) [X0]
+    + (-1.0) [X0 X1]
+    + (-0.5) [X0 Z1]
+    + (1.0) [Y0 Y1]
+    + (-0.5) [Z0 X1]
+    + (-0.5) [Z0 Y1]
```

Comment:
```suggestion
    (-1.0) [I<Wires = [0]> I<Wires = [1]>]
    + (-1.5) [I<Wires = [0]> X<Wires = [1]>]
    + (-0.5) [I<Wires = [0]> Y<Wires = [1]>]
    + (-1.0) [I<Wires = [0]> Z<Wires = [1]>]
    + (-1.5) [X<Wires = [0]> I<Wires = [1]>]
    + (-1.0) [X<Wires = [0]> X<Wires = [1]>]
    + (-0.5) [X<Wires = [0]> Z<Wires = [1]>]
    + (1.0) [Y<Wires = [0]> Y<Wires = [1]>]
    + (-0.5) [Z<Wires = [0]> X<Wires = [1]>]
    + (-0.5) [Z<Wires = [0]> Y<Wires = [1]>]
```
I think this is due to the new `Wires` class

Reply:
Note that this is a bug, and will be reverted in #684! So don't commit this suggestion, the original is how it should appear :)

Reply:
Good, the original is better :laughing: